File: stage3-translator/src/lexer/Lexer.java
package lexer;

import errors.SyntaxError;
import java.io.*;
import java.util.Hashtable;

public class Lexer {
	private char peek;
	private Hashtable<String, Token> words = new Hashtable<String, Token>();
	private InputStream input;
	public static int line = 1;

	private void addReserveWord(Word w) {
		words.put(w.getLexeme(), w);
	}

	public Lexer(InputStream input) {
		this.peek = ' ';

		this.input = input;
		addReserveWord(new Word("MAKE", Tag.MAKE));
		addReserveWord(new Word("FORWARD", Tag.FORWARD));
		addReserveWord(new Word("FD", Tag.FORWARD));
		addReserveWord(new Word("BACKWARD", Tag.BACKWARD));
		addReserveWord(new Word("BK", Tag.BACKWARD));
		addReserveWord(new Word("LEFT", Tag.LEFT));
		addReserveWord(new Word("LT", Tag.LEFT));
		addReserveWord(new Word("RIGHT", Tag.RIGHT));
		addReserveWord(new Word("RT", Tag.RIGHT));
		addReserveWord(new Word("SETX", Tag.SETX));
		addReserveWord(new Word("SETY", Tag.SETY));
		addReserveWord(new Word("SETXY", Tag.SETXY));
		addReserveWord(new Word("HOME", Tag.HOME));
		addReserveWord(new Word("CLEAR", Tag.CLEAR));
		addReserveWord(new Word("CLS", Tag.CLEAR));
		addReserveWord(new Word("CIRCLE", Tag.CIRCLE));
		addReserveWord(new Word("ARC", Tag.ARC));
		addReserveWord(new Word("PENUP", Tag.PENUP));
		addReserveWord(new Word("PU", Tag.PENUP));
		addReserveWord(new Word("PENDOWN", Tag.PENDOWN));
		addReserveWord(new Word("PD", Tag.PENDOWN));
		addReserveWord(new Word("COLOR", Tag.COLOR));
		addReserveWord(new Word("PENWIDTH", Tag.PENWIDTH));
		addReserveWord(new Word("PRINT", Tag.PRINT));
		addReserveWord(new Word("REPEAT", Tag.REPEAT));
		addReserveWord(new Word("IF", Tag.IF));
		addReserveWord(new Word("IFELSE", Tag.IFELSE));
		addReserveWord(new Word("NOT", Tag.NOT));
		addReserveWord(new Word("MOD", Tag.MOD));
		addReserveWord(new Word("AND", Tag.AND));
		addReserveWord(new Word("OR", Tag.OR));
	}

	private void readch() throws IOException {
		peek = (char) input.read();
	}

	private boolean readch(char c) throws IOException {
		readch();
		if (peek != c)
			return false;

		peek = ' ';
		return true;
	}

	private void skipSpaces() throws IOException {
		for ( ; ; readch() ) {
			if (peek == ' ' || peek == '\t' || peek == '\r') {
				continue;
			} else if (peek == '\n') {
				line = line + 1;
			} else {
				break;
			}
		}
	}

	public Token scan() throws IOException, SyntaxError {
		skipSpaces();

		if (peek == '%') {
			for ( ; ; readch() ) {
				if (peek == '\n') {
					line = line + 1;
					break;
				}
			}
			skipSpaces();
		}

		switch(peek) {
			case '<':
				if (readch('=')) return Word.Leq;
				else if (readch('>')) return Word.Neq;
				else return new Token('<');
			case '>':
				if (readch('=')) return Word.Geq;
				else return new Token('>');
			case '#':
				if (readch('t')) return Word.True;
				else if (readch('f')) return Word.False;
				else return new Token('#');
		}

		if (peek == '"') {
			String val = "";
			do {
				val = val + peek;
				readch();
			} while ( peek != '"' );
			val = val + peek;
			readch();
			return new Text(val);
		}

		if (Character.isDigit(peek)) {
			double number = 0;
			do {
				number = (10 * number) + Character.digit(peek, 10);
				readch();
			} while ( Character.isDigit(peek) );

			if (peek == '.') {
				readch();
				if (!Character.isDigit(peek)) {
					throw new SyntaxError("NUMERIC");
				}

				double divisor = 10.0;
				do {
					number = number + (Character.digit(peek, 10) / divisor);
					divisor = divisor * 10.0;
					readch();
				} while ( Character.isDigit(peek) );
			}
			return new Number(number);
		}

		if (Character.isLetter(peek)) {
			StringBuffer b = new StringBuffer();
			do {
				b.append(Character.toUpperCase(peek));
				readch();
			} while ( Character.isLetterOrDigit(peek) ) ;
			String s = b.toString();
			Word w = (Word) words.get(s);
			if (w != null)
				return w;

			w = new Word(s, Tag.ID);
			words.put(s, w);
			return w;
		}

		Token tok = new Token(peek); peek = ' ';
		return tok;
	}
}

File: stage-1/lexer.py
from enum import IntEnum

class Tag(IntEnum):
	EOF = 65535
	ERROR = 65534
	## Operators ##
	GEQ = 258
	LEQ = 259
	NEQ = 260
	ASSIGN = 261
	## REGULAR EXPRESSIONS ##
	ID = 357
	NUMBER = 358
	STRING = 359
	TRUE = 360
	FALSE = 361
	## RESERVED TokenS ##
	VAR = 457
	FORWARD = 458
	BACKWARD = 459
	LEFT = 460
	RIGHT = 461
	SETX = 462
	SETY = 463
	SETXY = 464
	CLEAR = 465
	CIRCLE = 466
	ARC = 467
	PENUP = 468
	PENDOWN = 469
	COLOR = 470
	PENWIDTH = 471
	PRINT = 472
	WHILE = 473
	IF = 474
	IFELSE = 475
	HOME = 476
	NOT = 477
	OR = 478
	AND = 479
	MOD = 480
	
class Token:
	__tag = Tag.EOF
	__value = None
	
	def __init__(self, tagId, val = None):
		self.__tag = tagId
		self.__value = val
		
	def getTag(self):
		return self.__tag
	
	def getValue(self):
		return self.__value
		
	def __str__(self):
		if self.__tag == Tag.GEQ:
			return "Token - value >="
		elif self.__tag == Tag.LEQ:
			return "Token - value <="
		elif self.__tag == Tag.NEQ:
			return "Token - value <>"
		elif self.__tag == Tag.ASSIGN:
			return "Token - value :="
		elif self.__tag == Tag.TRUE:
			return "Token - value TRUE"
		elif self.__tag == Tag.FALSE:
			return "Token - value FALSE"
		elif self.__tag == Tag.VAR:
			return "Token - value VAR"
		elif self.__tag == Tag.NUMBER:
			return "Number - value: " + str(self.__value)
		elif self.__tag == Tag.NUMBER:
			return "Number - value: " + str(self.__value)
		elif self.__tag == Tag.ID:
			return "Token - lexeme: " + str(self.__value)
		elif self.__tag >= Tag.VAR and self.__tag <= Tag.MOD:
			return "Reserved Token - lexeme: " + str(self.__value)
		elif self.__tag == Tag.STRING:
			return "String - text: " + str(self.__value)
		else:
			return "Token - value " + chr(self.__tag)
			
class Lexer:
	__peek = ' '
	__Tokens = {}
	__input = None
	__line = 1

	def __init__(self, filepath):
		self.__input = open(filepath, "r")
		self.__peek = ' '
		self.__line = 1

		self.__Tokens["VAR"] = Token(Tag.VAR, "VAR")
		self.__Tokens["FORWARD"] = Token(Tag.FORWARD, "FORWARD")
		self.__Tokens["FD"] = Token(Tag.FORWARD, "FORWARD")
		self.__Tokens["BACKWARD"] = Token(Tag.BACKWARD, "BACKWARD")
		self.__Tokens["BK"] = Token(Tag.BACKWARD, "BACKWARD")
		self.__Tokens["LEFT"] = Token(Tag.LEFT, "LEFT")
		self.__Tokens["LT"] = Token(Tag.LEFT, "LEFT")
		self.__Tokens["RIGHT"] = Token(Tag.RIGHT, "RIGHT")
		self.__Tokens["RT"] = Token(Tag.RIGHT, "RIGHT")
		self.__Tokens["SETX"] = Token(Tag.SETX, "SETX")
		self.__Tokens["SETY"] = Token(Tag.SETY, "SETY")
		self.__Tokens["SETXY"] = Token(Tag.SETXY, "SETXY")
		self.__Tokens["HOME"] = Token(Tag.HOME, "HOME")
		self.__Tokens["CLEAR"] = Token(Tag.CLEAR, "CLEAR")
		self.__Tokens["CLS"] = Token(Tag.CLEAR, "CLEAR")
		self.__Tokens["CIRCLE"] = Token(Tag.CIRCLE, "CIRCLE")
		self.__Tokens["ARC"] = Token(Tag.ARC, "ARC")
		self.__Tokens["PENUP"] = Token(Tag.PENUP, "PENUP")
		self.__Tokens["PU"] = Token(Tag.PENUP, "PENUP")
		self.__Tokens["PENDOWN"] = Token(Tag.PENDOWN, "PENDOWN")
		self.__Tokens["PD"] = Token(Tag.PENDOWN, "PENDOWN")
		self.__Tokens["COLOR"] = Token(Tag.COLOR, "COLOR")
		self.__Tokens["PENWIDTH"] = Token(Tag.PENWIDTH, "PENWIDTH")
		self.__Tokens["PRINT"] = Token(Tag.PRINT, "PRINT")
		self.__Tokens["WHILE"] = Token(Tag.WHILE, "WHILE")
		self.__Tokens["IF"] = Token(Tag.IF, "IF")
		self.__Tokens["IFELSE"] = Token(Tag.IFELSE, "IFELSE")
		self.__Tokens["NOT"] = Token(Tag.NOT, "NOT")
		self.__Tokens["OR"] = Token(Tag.OR, "OR")
		self.__Tokens["AND"] = Token(Tag.AND, "AND")
		self.__Tokens["MOD"] = Token(Tag.MOD, "MOD")

	def read(self):
		self.__peek = self.__input.read(1)
	
	def readch(self, c):
		self.read()
		if self.__peek != c:
			return False

		self.__peek = ' '
		return True

	def __skipSpaces(self):
		while True:
			if self.__peek == ' ' or self.__peek == '\t' or self.__peek == '\r':
				self.read()
			elif self.__peek == '\n':
				self.read()
				self.__line = self.__line + 1
			else:
				break
	
	def scan(self):
		self.__skipSpaces()

		## ADD CODE TO SKIP COMMENTS HERE ##

		if self.__peek == '<':
			if self.readch('='):
				return Token(Tag.LEQ, "<=")
			elif self.readch('>'):
				return Token(Tag.NEQ, "<>")
			else:
				return Token(ord('<'))
		elif self.__peek == '>':
			if self.readch('='):
				return Token(Tag.GEQ, ">=")
			else:
				return Token(ord('>'))
		elif self.__peek == '#':
			if self.readch('t'):
				return Token(Tag.TRUE, "#t")
			elif self.readch('>'):
				return Token(Tag.FALSE, "#f")
			else:
				return Token(ord('#'))
		elif self.__peek == ':':
			if self.readch('='):
				return Token(Tag.ASSIGN, ":=")
			else:
				return Token(ord(':'))

		if self.__peek  == '"':
			val = ""
			while True:
				val = val + self.__peek
				self.read()
				if self.__peek == '"':
					break
			
			val = val + self.__peek
			self.read()
			return Token(Tag.STRING, val)

		if self.__peek.isdigit():
			val = 0
			while True:
				val = (val * 10) + int(self.__peek)
				self.read()
				if not(self.__peek.isdigit()):
					break
			## ADD CODE TO PROCESS DECIMAL PART HERE ##
			return Token(Tag.NUMBER, val)

		if self.__peek.isalpha():
			val = ""
			while True:
				val = val + self.__peek.upper()
				self.read()
				if not(self.__peek.isalnum()):
					break

			if val in self.__Tokens:
				return self.__Tokens[val]

			w = Token(Tag.ID, val)
			self.__Tokens[val] = Token(Tag.ID, val)
			return w

		if not(self.__peek):
			return Token(Tag.EOF)			

		token = Token(ord(self.__peek))
		self.__peek = ' ' 
		return token
